README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# meantables

<!-- badges: start -->

[![Travis build
status](https://travis-ci.com/brad-cannell/meantables.svg?branch=master)](https://travis-ci.com/brad-cannell/meantables)
[![CRAN
status](https://www.r-pkg.org/badges/version/meantables)](https://cran.r-project.org/package=meantables)
<!-- badges: end -->

The goal of meantables is to quickly make tables of descriptive
statistics (i.e., counts, means, confidence intervals) for continuous
variables. This package is designed to work in a Tidyverse pipeline, and
consideration has been given to get results from R to ‘Microsoft Word’ ®
with minimal pain.

## Installation

You can install the released version of meantables from
[CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("meantables")
```

And the development version from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("brad-cannell/meantables")
```

## Example

``` r
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
library(meantables)
```

``` r
data("mtcars")
```

### Overall mean table with defaults

``` r
mtcars %>% 
  mean_table(mpg)
#> # A tibble: 1 × 9
#>   response_var     n  mean    sd   sem   lcl   ucl   min   max
#>   <chr>        <int> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl>
#> 1 mpg             32  20.1  6.03  1.07  17.9  22.3  10.4  33.9
```

### Formatting overall mean and 95% CI

``` r
mtcars %>%
  mean_table(mpg) %>%
  mean_format(
    recipe = "mean (lcl - ucl)",
    name = "mean_95",
    digits = 1
  ) %>% 
  select(response_var, mean_95)
#> # A tibble: 1 × 2
#>   response_var mean_95           
#>   <chr>        <chr>             
#> 1 mpg          20.1 (17.9 - 22.3)
```

### Formatting grouped means table with mean and sd

``` r
mtcars %>%
  group_by(cyl) %>%
  mean_table(mpg) %>%
  mean_format("mean (sd)") %>% 
  select(response_var:group_cat, formatted_stats)
#> # A tibble: 3 × 4
#>   response_var group_var group_cat formatted_stats
#>   <chr>        <chr>         <dbl> <chr>          
#> 1 mpg          cyl               4 26.66 (4.51)   
#> 2 mpg          cyl               6 19.74 (1.45)   
#> 3 mpg          cyl               8 15.1 (2.56)
```

### Grouped means table with defaults

``` r
mtcars %>% 
  group_by(cyl) %>% 
  mean_table(mpg)
#> # A tibble: 3 × 11
#>   response_var group_var group_cat     n  mean    sd   sem   lcl   ucl   min
#>   <chr>        <chr>         <dbl> <int> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl>
#> 1 mpg          cyl               4    11  26.7  4.51 1.36   23.6  29.7  21.4
#> 2 mpg          cyl               6     7  19.7  1.45 0.549  18.4  21.1  17.8
#> 3 mpg          cyl               8    14  15.1  2.56 0.684  13.6  16.6  10.4
#> # … with 1 more variable: max <dbl>
```


NEWS.md
# meantables 0.1.2

* Added the `mean_format()` function.

# meantables 0.1.1

* Changed the way that `n_miss` and `n` are calculated. Now correct when there are missing values. Built unit test to check for this.

* Changed the `x` argument of `mean_table()` to `.x`. This prevents an error when passing a variable named `x` to `mean_table()`.

* Added standard deviation to `mean_table()` results.

# meantables 0.1.0

* First release on CRAN


tests/testthat.R
library(testthat)
library(meantables)

test_check("meantables")


tests/testthat/test-mean_table.R
library(dplyr)
library(meantables)

data(mtcars)

testthat::context("test-mean_table.R")

# =============================================================================
# N is calculated correctly when there are missing values
# =============================================================================
df <- tibble(
  x = c(1, 2, NA, 4, 5)
)

df <- df %>%
  mean_table(x, output = "all")

testthat::test_that("N is calculated correctly when there are missing values", {
  testthat::expect_equal(df$n_miss, 1L)
  testthat::expect_equal(df$n, 4L)
})

# =============================================================================
# Test one-way mean tables
# =============================================================================
df <- mtcars %>%
  mean_table(mpg)

testthat::test_that("Dimensions of the object returned by mean_table are as expected", {
  rows    <- nrow(df)
  columns <- ncol(df)

  testthat::expect_equal(rows, 1L)
  testthat::expect_equal(columns, 9L)
})

testthat::test_that("Class of mean_table_one_way is mean_table", {
  testthat::expect_is(df, "mean_table")
})

testthat::test_that("The correct var name is returned by mean_table", {
  name <- names(df)[1]
  testthat::expect_match(name, "response_var")
})

testthat::test_that("The correct default statistics are returned by mean_table", {
  n    <- pull(df, n)
  mean <- pull(df, mean)
  sd   <- pull(df, sd)
  sem  <- pull(df, sem) %>% round(2)
  lcl  <- pull(df, lcl) %>% round(2)
  ucl  <- pull(df, ucl) %>% round(2)
  min  <- pull(df, min)
  max  <- pull(df, max)

  testthat::expect_equal(n, 32)
  testthat::expect_equal(mean, 20.09)
  testthat::expect_equal(sd, 6.03)
  testthat::expect_equal(sem, 1.07)
  testthat::expect_equal(lcl, 17.92)
  testthat::expect_equal(ucl, 22.26)
  testthat::expect_equal(min, 10.4)
  testthat::expect_equal(max, 33.9)
})


# =============================================================================
# Test grouped mean tables
# =============================================================================
df <- mtcars %>%
  group_by(cyl) %>%
  mean_table(mpg)

testthat::test_that("Dimensions of the object returned by mean_table are as expected", {
  rows    <- nrow(df)
  columns <- ncol(df)

  testthat::expect_equal(rows, 3L)
  testthat::expect_equal(columns, 11L)
})

testthat::test_that("Class of mean_table_two_way is mean_table_grouped", {
  testthat::expect_is(df, "mean_table_grouped")
})

testthat::test_that("The correct var names are returned by mean_table", {
  row_var <- pull(df, response_var)
  col_var <- pull(df, group_var)

  testthat::expect_match(row_var, "mpg")
  testthat::expect_match(col_var, "cyl")
})

testthat::test_that("The correct variables levels are returned by mean_table", {
  group_cat <- pull(df, group_cat)

  testthat::expect_equal(group_cat, c(4, 6, 8))
})

testthat::test_that("The correct default statistics are returned by mean_table", {
  n    <- pull(df, n)
  mean <- pull(df, mean)
  sd   <- pull(df, sd)
  sem  <- pull(df, sem) %>% round(2)
  lcl  <- pull(df, lcl) %>% round(2)
  ucl  <- pull(df, ucl) %>% round(2)
  min  <- pull(df, min)
  max  <- pull(df, max)

  testthat::expect_equal(n, c(11, 7, 14))
  testthat::expect_equal(mean, c(26.66, 19.74, 15.10))
  testthat::expect_equal(sd, c(4.51, 1.45, 2.56))
  testthat::expect_equal(sem, c(1.36, 0.55, 0.68))
  testthat::expect_equal(lcl, c(23.63, 18.40, 13.62))
  testthat::expect_equal(ucl, c(29.69, 21.09, 16.58))
  testthat::expect_equal(min, c(21.4, 17.8, 10.4))
  testthat::expect_equal(max, c(33.9, 21.4, 19.2))
})


# =============================================================================
# Clean up
# =============================================================================
rm(mtcars, df)
detach("package:dplyr", unload=TRUE)
detach("package:meantables", unload=TRUE)


tests/testthat/test-mean_format.R
library(dplyr)
library(meantables)

data(mtcars)

testthat::context("test-mean_format.R")

# =============================================================================
# Test ungrouped mean table, stats = mean and 95% CI
# =============================================================================
df <- mtcars %>%
  mean_table(mpg) %>%
  mean_format(
    recipe = "mean (lcl - ucl)",
    name = "mean_95",
    digits = 1
  )

testthat::test_that("Dimensions of the object returned by mean_format are as expected", {
  dims <- dim(df)
  testthat::expect_equal(dims, c(1L, 10L))
})

testthat::test_that("The the name argument to mean_format works as expected", {
  name <- names(df)[10]
  testthat::expect_match(name, "mean_95")
})

testthat::test_that("The correct statistics are returned by mean_format", {
  mean_95 <- pull(df, mean_95)
  testthat::expect_equal(mean_95, "20.1 (17.9 - 22.3)")
})


# =============================================================================
# Test grouped means table, stats = mean and sd
# =============================================================================
df <- mtcars %>%
  group_by(cyl) %>%
  mean_table(mpg) %>%
  mean_format("mean (sd)")

testthat::test_that("Dimensions of the object returned by mean_format are as expected", {
  dims <- dim(df)
  testthat::expect_equal(dims, c(3L, 12L))
})

testthat::test_that("The correct statistics are returned by mean_format", {
  mean_sd <- pull(df, formatted_stats)
  testthat::expect_equal(mean_sd, c("26.66 (4.51)", "19.74 (1.45)", "15.1 (2.56)"))
})


# =============================================================================
# Clean up
# =============================================================================
rm(mtcars, df)
detach("package:dplyr", unload=TRUE)
detach("package:meantables", unload=TRUE)


R/mean_format.R
#' @title Format mean_table Output for Publication and Dissemination
#'
#' @description The mean_format function is intended to make it quick and easy to
#'   format the output of the mean_table function for tables that may be used
#'   for publication. For example, a mean and 95% confidence interval
#'   could be formatted as "24.00 (21.00 - 27.00)."
#'
#' @param .data A data frame of class "mean_table" or "mean_table_grouped".
#'
#' @param recipe A recipe used to create a new column from existing mean_table
#'   columns. The recipe must be in the form of a quoted string. It may contain
#'   any combination of column names, spaces, and characters. For example:
#'   "mean (sd)" or "mean (lcl - ucl)".
#'
#' @param name An optional name to assign to the column created by the recipe.
#'   The default name is "formatted_stats"
#'
#' @param digits The number of decimal places to display.
#'
#' @return A tibble
#' @export
#'
#' @examples
#' \dontrun{
#' library(dplyr)
#' library(meantables)
#'
#' data(mtcars)
#'
#' # Overall mean table with defaults
#'
#' mtcars %>%
#'   mean_table(mpg) %>%
#'   mean_format("mean (sd)") %>%
#'   select(response_var, formatted_stats)
#'
#' # A tibble: 1 × 2
#'   response_var formatted_stats
#'   <chr>        <chr>
#' 1 mpg          20.09 (6.03)
#'
#' # Grouped means table with defaults
#'
#' mtcars %>%
#'   group_by(cyl) %>%
#'   mean_table(mpg) %>%
#'   mean_format("mean (sd)") %>%
#'   select(response_var:group_cat, formatted_stats)
#'
#'   # A tibble: 3 × 4
#'   response_var group_var group_cat formatted_stats
#'   <chr>        <chr>         <dbl> <chr>
#' 1 mpg          cyl               4 26.66 (4.51)
#' 2 mpg          cyl               6 19.74 (1.45)
#' 3 mpg          cyl               8 15.1 (2.56)
#' }
mean_format <- function(.data, recipe, name = NA, digits = NA) {

  # ===========================================================================
  # Prevents R CMD check: "no visible binding for global variable ‘.’"
  # ===========================================================================
  # name = recipe = ingredients = stat = NULL

  # ===========================================================================
  # Check function arguments
  # ===========================================================================
  # If no name given, default to formatted_stats
  if(is.na(name)) {
    name <- "formatted_stats"
  }

  # Break up the recipe into its component pieces
  # [1] "" "n" " (" "percent" ")"
  recipe <- stringr::str_split(recipe, "\\b")
  # First component is always an empty string. Drop it.
  recipe <- unlist(recipe)[-1]

  # Loop over each row of the mean_table
  for(i in seq(nrow(.data))) {
    # Empty vector to hold the stats and symbols that will make up the new
    # variable in that row. The ingredients for the recipe.
    ingredients <- c()
    # Loop over each component of the recipe
    # [1] "n" " (" "percent" ")"
    for(j in seq_along(recipe)) {
      # If that the component is a column in the data frame then grab the value
      # for that column in that row and add it to the ingredients.
      if(recipe[j] %in% names(.data)){
        # Get the stat (e.g., n or percent)
        stat <- .data[[recipe[j]]][i]
        # Round the stat if an argument is supplied to the digits argument
        if(!is.na(digits)) {
          # But don't add trailing zeros to integers
          if(!is.integer(stat)) {
            stat <- round(stat, digits)
            stat <- format(stat, nsmall = digits, big.mark = ",")
            stat <- trimws(stat)
          }
        }
        ingredients <- c(ingredients, stat)
        # If that component is not a column in .data then add it to the
        # ingredients vector as a character.
      } else {
        ingredients <- c(ingredients, recipe[j])
      }
    }
    # Add the new variable to the .data
    .data[i, name] <- paste(ingredients, collapse = "")
  }

  # Return .data with the new variable
  .data
}


R/mean_table.R
#' @title Estimate Mean and 95 Percent Confidence Intervals in dplyr Pipelines
#'
#' @description The mean_table function produces overall and grouped
#'   tables of means with related statistics. In addition to means, the
#'   mean_table missing/non-missing frequencies, the standard error of the
#'   mean (sem), the 95% confidence intervals for the mean(s), the minimum
#'   value, and the maximum value. For grouped tibbles, mean_table displays
#'   these statistics for each category of the group_by variable.
#'
#' @param .data A tibble or grouped tibble.
#'
#' @param .x The continuous response variable for which the statistics are
#'   desired.
#'
#' @param t_prob (1 - alpha / 2). Default value is 0.975, which corresponds to
#'   an alpha of 0.05. Used to calculate a critical value from Student's t
#'   distribution with n - 1 degrees of freedom.
#'
#' @param output Options for this parameter are "default" and "all".
#'
#'   Default output includes the n, mean, sem, and 95% confidence interval for
#'   the mean. Using output = "all" also returns the the number of missing
#'   values for .x and the critical t-value.
#'
#' @param digits Round mean, lcl, and ucl to digits. Default is 2.
#'
#' @param ... Other parameters to be passed on.
#'
#' @return A tibble of class "mean_table" or "mean_table_grouped"
#' @export
#' @importFrom dplyr %>%
#'
#' @references
#'   SAS documentation: http://support.sas.com/documentation/cdl/en/proc/65145/HTML/default/viewer.htm#p0klmrp4k89pz0n1p72t0clpavyx.htm
#'
#' @examples
#' \dontrun{
#' library(dplyr)
#' library(meantables)
#'
#' data(mtcars)
#'
#' # Overall mean table with defaults
#'
#' mtcars %>%
#'   mean_table(mpg)
#'
#' # A tibble: 1 x 9
#'   response_var     n  mean    sd   sem   lcl   ucl   min   max
#'   <chr>        <int> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl>
#' 1 mpg             32  20.1  6.03  1.07  17.9  22.3  10.4  33.9
#'
#' # Grouped means table with defaults
#'
#' mtcars %>%
#'   group_by(cyl) %>%
#'   mean_table(mpg)
#'
#' # A tibble: 3 x 11
#'   response_var group_var group_cat     n  mean    sd   sem   lcl   ucl   min   max
#'   <chr>        <chr>         <dbl> <int> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl>
#' 1 mpg          cyl               4    11  26.7  4.51 1.36   23.6  29.7  21.4  33.9
#' 2 mpg          cyl               6     7  19.7  1.45 0.549  18.4  21.1  17.8  21.4
#' 3 mpg          cyl               8    14  15.1  2.56 0.684  13.6  16.6  10.4  19.2
#' }

mean_table <- function(.data, .x, t_prob = 0.975, output = default, digits = 2, ...) {

  # ------------------------------------------------------------------
  # Prevents R CMD check: "no visible binding for global variable ‘.’"
  # ------------------------------------------------------------------
  n = sd = t_crit = sem = lcl = ucl = var = n_groups = group_1 = NULL
  group_2 = output_arg = default = response_var = `.` = NULL
  group_var = group_cat = group_1_cat = group_2_cat = n_miss = NULL

  # ===========================================================================
  # Quick data checks
  # Input data frame is class data.frame
  # The ".x" argument is a numeric vector
  # There are zero, one, or two group_by variables
  # ===========================================================================
  if (!("data.frame" %in% class(.data))) {
    message("Expecting the class of .data to include data.frame. Instead, the ",
            "class was ", class(.data))
  }

  if (missing(.x)) {
    stop("No argument was passed to the '.x' parameter. Expecting '.x' to be a ",
         "numeric column.")
  }

  # ===========================================================================
  # Enquo arguments
  # enquo the .x argument so that it can be used in the dplyr pipeline below.
  # The .x argument is the variable you want the mean of.
  # enquo/quo_name/UQ the output argument so that I don't have to use
  # quotation marks around the argument being passed.
  # ===========================================================================
  .x          <- rlang::enquo(.x)
  output_arg <- rlang::enquo(output) %>% rlang::quo_name()

  # ===========================================================================
  # Grouping variables
  # Count the number of them - accept zero, one, or two
  # Grab their names - later returned in summary table
  # ===========================================================================
  if ("grouped_df" %in% class(.data)) {
    n_groups <- attributes(.data)$groups %>% length() - 1
  } else {
    n_groups <- 0L
  }

  if (n_groups == 1) {
    group_1 <- attributes(.data)$groups[1] %>% names()

  } else if (n_groups == 2) {
    group_1 <- attributes(.data)$groups[1] %>% names()
    group_2 <- attributes(.data)$groups[2] %>% names()

  } else if (n_groups > 2) {
    stop(".data can be grouped by up to two variables. It is currently grouped ",
         n_groups, " variables")
  }

  # ===========================================================================
  # Create a general summary table of means and related stats. Then, add
  # group variable names to summary table where applicable
  # 1. No group_by variables
  # 2. One group_by variable
  # 3. Two group_by variables
  # ===========================================================================
  out <- .data %>%
    dplyr::summarise(
      # Grab variable (.x) name
      response_var = rlang::quo_name(.x),
      n_miss       = sum(is.na(.data[[rlang::quo_name(.x)]])),
      n            = sum(!is.na(.data[[rlang::quo_name(.x)]])),
      mean         = mean(!! .x, na.rm = TRUE),
      sd           = stats::sd(!! .x, na.rm = TRUE),
      t_crit       = stats::qt(t_prob, n - 1),
      sem          = sd / sqrt(n),
      lcl          = mean - t_crit * sem,
      ucl          = mean + t_crit * sem,
      mean         = round(mean, digits),   # Round mean
      sd           = round(sd, digits),     # Round sd
      lcl          = round(lcl, digits),    # Round lcl
      ucl          = round(ucl, digits),    # Round ucl
      min          = min(!! .x, na.rm = TRUE),
      max          = max(!! .x, na.rm = TRUE)
    ) %>%
    tibble::as_tibble()

  # ===========================================================================
  # Add group variable names to summary table - if applicable
  # Then move to the front of the summary table.

  # Also add classes to "out"
  # If the input data frame (.data) was a grouped data frame, then the output
  # will be a bivariate analysis of means ("mean_table_grouped"). Pass that
  # information on to "out." It can be used later in mean_format.
  # Otherwise the output will be a univariate analysis of means ("mean_table")
  # That class will also be used later in mean_format.
  # ===========================================================================
  if (n_groups == 0) {

    out <- out %>%
      dplyr::select(response_var, dplyr::everything())
    class(out) <- c("mean_table", class(out))

  }
  else if (n_groups == 1) {

    out <- out %>%
      dplyr::mutate(group_var = group_1) %>%
      dplyr::rename(group_cat = !! names(.)[1]) %>%
      dplyr::select(response_var, group_var, group_cat, dplyr::everything())
    class(out) <- c("mean_table_grouped", class(out))

  } else if (n_groups == 2) {

    out <- out %>%
      dplyr::mutate(
        group_1 = group_1,
        group_2 = group_2
      ) %>%
      dplyr::rename(
        group_1_cat = !! names(.)[1],
        group_2_cat = !! names(.)[2]
      ) %>%
      dplyr::select(response_var, group_1, group_1_cat, group_2, group_2_cat,
                    dplyr::everything()) %>%
      dplyr::ungroup()
    class(out) <- c("mean_table_grouped", class(out))
  }

  # ===========================================================================
  # Control output:
  # Typically, I only want the frequency, mean, 95% CI, sem, min, and max.
  # Make that the default.
  # ===========================================================================
  if (output_arg == "default") {
    out <- out %>%
      dplyr::select(-c(n_miss, t_crit))
  }

  # Return summary table
  out
}
